libraries/AP_NavEKF3/AP_NavEKF3.cpp

```diff
@@ -815,34 +821,53 @@ void NavEKF3::UpdateFilter(void)
         }
         runCoreSelection = (imuSampleTime_us - lastUnhealthyTime_us) > 1E7;
     }
-    float primaryErrorScore = core[primary].errorScore();
-    if ((primaryErrorScore > 1.0f || !core[primary].healthy()) && runCoreSelection) {
-        float lowestErrorScore = 0.67f * primaryErrorScore;
-        uint8_t newPrimaryIndex = primary; // index for new primary
-        for (uint8_t coreIndex=0; coreIndex<num_cores; coreIndex++) {
+    
+                                   
+    if (runCoreSelection) {
+        float primaryErrorScore = core[primary].errorScore();
+
+        // update the relative error scores for all active cores
+        updateCoreErrors(primaryErrorScore);
 
+        bool betterCore = false;
+        bool altCoreAvailable = false;
+        float maxCoreError = relativeCoreError[primary]; // looking for cores that have error lower than the current primary
+        uint8_t newPrimaryIndex = primary;
+
+        // loop through all available cores to find if an alternative core is available
+        for (uint8_t coreIndex=0; coreIndex<num_cores; coreIndex++) {
             if (coreIndex != primary) {
-                // an alternative core is available for selection only if healthy and if states have been updated on this time step
-                bool altCoreAvailable = core[coreIndex].healthy() && statePredictEnabled[coreIndex];
-
-                // If the primary core is unhealthy and another core is available, then switch now
-                // If the primary core is still healthy,then switching is optional and will only be done if
-                // a core with a significantly lower error score can be found
-                float altErrorScore = core[coreIndex].errorScore();
-                if (altCoreAvailable && (!core[newPrimaryIndex].healthy() || altErrorScore < lowestErrorScore)) {
+                float altCoreError = relativeCoreError[coreIndex];
+
+                // an alternative core is available for selection based on 2 conditions -
+                // 1. healthy and states have been updated on this time step
+                // 2. has relative error less than primary core error
+                altCoreAvailable = core[coreIndex].healthy() && statePredictEnabled[coreIndex] && altCoreError < maxCoreError;
+
+                if(altCoreAvailable) {
+                    // if this core has a significantly lower relative error to the active primary, we consider it as a 
+                    // better core and would like to switch to it even if the current primary is healthy
+                    betterCore = (0.5*CORE_ERR_LIM + altCoreError) < relativeCoreError[primary]; // consider a better core if its relative error is lower by half the max possible error
```

Comment:
I think the 0.5*CORE_ERR_LIM should be replaced by a separate threshold define. The CORE_ERR_LIM and this threshold are not really related.